exchanges/kucoin/kucoin_websocket.go

```diff
@@ -0,0 +1,1679 @@
+	_, err := ku.GetCredentials(context.Background())
+	if err != nil {
+		ku.Websocket.SetCanUseAuthenticatedEndpoints(false)
+	}
+	if ku.Websocket.CanUseAuthenticatedEndpoints() {
+		instances, _ = ku.GetAuthenticatedInstanceServers(context.Background())
+		if err != nil {
+			ku.Websocket.DataHandler <- err
+			ku.Websocket.SetCanUseAuthenticatedEndpoints(false)
+		}
+	}
+	if instances == nil {
+		instances, err = ku.GetInstanceServers(context.Background())
+		if err != nil {
+			return err
+		}
+	}
+	if len(instances.InstanceServers) == 0 {
+		return errors.New("no websocket instance server found")
+	}
+	ku.Websocket.Conn.SetURL(instances.InstanceServers[0].Endpoint + "?token=" + instances.Token)
+	err = ku.Websocket.Conn.Dial(&dialer, http.Header{})
+	if err != nil {
+		return fmt.Errorf("%v - Unable to connect to Websocket. Error: %s", ku.Name, err)
+	}
+	ku.Websocket.Wg.Add(1)
+	go ku.wsReadData()
+	if err != nil {
+		return err
+	}
+	ku.Websocket.Conn.SetupPingHandler(stream.PingHandler{
+		Delay:       time.Millisecond * time.Duration(instances.InstanceServers[0].PingTimeout),
+		Message:     []byte(`{"type":"ping"}`),
+		MessageType: websocket.TextMessage,
+	})
+
+	ku.setupOrderbookManager()
+	return nil
+}
+
+// GetInstanceServers retrieves the server list and temporary public token
+func (ku *Kucoin) GetInstanceServers(ctx context.Context) (*WSInstanceServers, error) {
+	response := struct {
+		Data WSInstanceServers `json:"data"`
+		Error
+	}{}
+	return &(response.Data), ku.SendPayload(ctx, request.Unset, func() (*request.Item, error) {
+		endpointPath, err := ku.API.Endpoints.GetURL(exchange.RestSpot)
+		if err != nil {
+			return nil, err
+		}
+		return &request.Item{
+			Method:        http.MethodPost,
+			Path:          endpointPath + publicBullets,
+			Result:        &response,
+			AuthRequest:   true,
+			Verbose:       ku.Verbose,
+			HTTPDebugging: ku.HTTPDebugging,
+			HTTPRecording: ku.HTTPRecording}, nil
+	})
+}
+
+// GetAuthenticatedInstanceServers retrieves server instances for authenticated users.
+func (ku *Kucoin) GetAuthenticatedInstanceServers(ctx context.Context) (*WSInstanceServers, error) {
+	response := struct {
+		Data *WSInstanceServers `json:"data"`
+		Error
+	}{}
+	err := ku.SendAuthHTTPRequest(ctx, exchange.RestSpot, defaultSpotEPL, http.MethodPost, privateBullets, nil, &response)
+	if err != nil && strings.Contains(err.Error(), "400003") {
+		return response.Data, ku.SendAuthHTTPRequest(ctx, exchange.RestFutures, defaultFuturesEPL, http.MethodPost, privateBullets, nil, &response)
+	}
+	return response.Data, err
+}
+
+// wsReadData receives and passes on websocket messages for processing
+func (ku *Kucoin) wsReadData() {
+	defer ku.Websocket.Wg.Done()
+	for {
+		resp := ku.Websocket.Conn.ReadMessage()
+		if resp.Raw == nil {
+			return
+		}
+		err := ku.wsHandleData(resp.Raw)
+		if err != nil {
+			ku.Websocket.DataHandler <- err
+		}
+	}
+}
+
+func (ku *Kucoin) wsHandleData(respData []byte) error {
```

Comment:
Looks like there's an incorrect assumption for storing values:
[ERROR] | WEBSOCKET | 20/02/2023 14:50:28 | exchange Kucoin websocket error - unsupported integer value

Likely related to this:
[INFO]  | SYNC | 20/02/2023 14:50:28 | Kucoin websocket btcusdt SPOT TICKER: Last 0.00026058 Ask 24402.90000000 Bid 24402.80000000 High 0.00000000 Low 0.00000000 Volume 0.00000000
But could be seperate, in which case, please fix the above too. No way there is zero volume on BTC